### ISSUE TITLE
Typed array instance used as prototype incorrect perform [[Get]]/[[Set]]

### ISSUE THREAD

**Issue by allenwb** [tc39/ecma262#1541]
Using a Typed Array (or most any exotic object) as as a prototype is weird.  But their specification
 of [\[\[Get\]\]](https://tc39.github.io/ecma262/#sec-integer-indexed-exotic-objects-get-p-receiver) and [\[\[Set\]\]](https://tc39.github.io/ecma262/#sec-integer-indexed-exotic-objects-set-p-v-receiver) does not correctly deal with the case where the _Rceiver_ argument is not the same object as the exotic array.  In this case, a [[Put]] can actually mutate the prototype type object.   Consider:

```js
let o = new Int8Array(1)
Object.setPrototypeOf(Array.prototype, o);
const a = [];  //a inherits indirectly from o
a[0] = 4;  //according to current spec. this will set o[0] to 4.
```

The problem is that [9.4.5.5](https://tc39.github.io/ecma262/#sec-integer-indexed-exotic-objects-set-p-v-receiver) does not check if _O_ is different from _Receiver_ and calls IntegerIndexElementSet to mutate _O_ rather than _Receiver_ (which would also be wrong, but for other reasons.

The fix is to change step 2 to:

> if SameValue(_O_,_Reciever_) and Type(_P_) is String, then

The same change should also be made to [9.4.5.4](https://tc39.github.io/ecma262/#sec-integer-indexed-exotic-objects-get-p-receiver)

**Comment by ljharb** [comment]
With this code:
```js
let o = new Int8Array(1)
Object.setPrototypeOf(Array.prototype, o);
const a = [];  //a inherits indirectly from o
a[0] = 4;  //according to current spec. this will set o[0] to 4.
print(a[0], o[0]);
```
I get these results from `eshost file.js`:
```sh
#### Chakra
4 0

#### JavaScriptCore
4 0

#### SpiderMonkey
4 0

#### V8
4 0

#### V8 --harmony
4 0
```

This suggests that engines are already doing the correct thing, so it's just a bug in the spec.

**Comment by anba** [comment]
That means reverting #347, so we'd again get what was originally specified in ES6.

**Comment by ljharb** [comment]
Does that imply that engines implemented it after #347 landed but with the prior behavior?

**Comment by anba** [comment]
I guess SpiderMonkey follows what the current spec (including #347) has for [9.4.5.4 [[Get]]](https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects-get-p-receiver), but SpiderMonkey's implementation for [9.4.5.5 [[Set]]](https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects-set-p-v-receiver) doesn't seem to comply to the current spec.

[LookupOwnPropertyInline](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/js/src/vm/NativeObject-inl.h#743,766) returns `done = true` when a TypedArray object is encountered, so we always stop in [NativeGetPropertyInline](https://searchfox.org/mozilla-central/source/js/src/vm/NativeObject.cpp#2544,2561,2581,2585) even for out-of-bounds accesses. But for [[Set]], an out-of-bounds access results in adding a new property in [NativeSetProperty](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/js/src/vm/NativeObject.cpp#2940,2955,2972-2977) and also in [SetExistingProperty](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/js/src/vm/NativeObject.cpp#2886,2898) for in-bounds, but wrong receiver.

So basically the primary reason for any differences in SpiderMonkey's implementation, is that we inspect the [[ArrayLength]] internal slot of the TypedArrray object [earlier](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/js/src/vm/NativeObject-inl.h#761-765) than specified in the spec.

**Comment by ljharb** [comment]
OK, so I guess the real questions here imo are in this order:
 1. what behavior do we *want*?
 1. are engines willing to change to the desired behavior?
 1. if so, yay, let's merge that! if not, what's the most desirable behavior that engines are willing to implement, and how does that differ from current reality?

**Comment by evilpie** [comment]
I think I tried to implement the spec behavior for [[Set]] and it's not web compatible: https://bugzilla.mozilla.org/show_bug.cgi?id=1502889

**Comment by ljharb** [comment]
aha, and that points to #1339.

Looks like we may need to revert all or part of #347?

**Comment by anba** [comment]
I've totally forgotten about the web-compatibility issue encountered in Firefox.

> Looks like we may need to revert all or part of #347?

I guess the [[Get]] part is still okay, but [[Set]] will need to be changed.

And then we need to decide if consistency between [[Get]] and [[Set]] (*) or between [[Get]] and [[HasProperty]] (**) is more important, because we can't get both.

(*) That means consistent `receiver` checks.
(**) Per the example in #347

**Comment by verwaest** [comment]
What about just adding a branch for into 2 (if SameValue(O,Receiver)) so we don't fall to 3, but still don't set the value onto the prototype?

**Comment by evilpie** [comment]
@rkirsling this might be relevant to you, considering that you worked on #2164.

**Comment by shvaikalesh** [comment]
@evilpie I've updated #1556 to align the spec with V8, fixing the web-compatibility issue that is still present. I will update the summary and submit thorough test262 coverage within few days.

@rkirsling would you please consider presenting it at the next TC39 meeting?

**Comment by syg** [comment]
I got consensus for [this](https://docs.google.com/presentation/d/1Cg7XDFmq2_Aa7h1ZNdgrmT_UCIm0w02P9Z7Krht46Kw/edit#slide=id.p) back in Dec 2019, but embarrassingly forgot to ever write up a normative PR for it. Is this the same issue or a different one?

**Comment by evilpie** [comment]
> I got consensus for [this](https://docs.google.com/presentation/d/1Cg7XDFmq2_Aa7h1ZNdgrmT_UCIm0w02P9Z7Krht46Kw/edit#slide=id.p) back in Dec 2019, but embarrassingly forgot to ever write up a normative PR for it. Is this the same issue or a different one?

Oh nice! I am pretty sure the [[Set]] issue is the same. You are even showing the Vector example that I found.

**Comment by akamsmrhill** [comment]
yes

**Comment by shvaikalesh** [comment]
@syg It is definitely the same issue, yet with a bit different solution for `[[Set]]`. [Slide 10](https://docs.google.com/presentation/d/1Cg7XDFmq2_Aa7h1ZNdgrmT_UCIm0w02P9Z7Krht46Kw/edit#slide=id.g75ac69d4fb_0_121) claims:

> Inside IntegerIndexed when the receiver is not the holder inside TypedArrays’ [[Set]], do an OrdinarySet.

Taking into account [Slide 12](https://docs.google.com/presentation/d/1Cg7XDFmq2_Aa7h1ZNdgrmT_UCIm0w02P9Z7Krht46Kw/edit#slide=id.g75ac69d4fb_0_126), it translates to normative change like:

```diff
          1. Assert: IsPropertyKey(_P_) is *true*.
-         1. If Type(_P_) is String, then
+         1. If SameValue(_O_, _Receiver_) is *true* and Type(_P_) is String, then
            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
            1. If _numericIndex_ is not *undefined*, then
              1. Perform ? IntegerIndexedElementSet(_O_, _numericIndex_, _V_).
              1. Return *true*.
          1. Return ? OrdinarySet(_O_, _P_, _V_, _Receiver_).
```

Which would break consistency with TypedArray's [`[[HasProperty]]`](https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects-hasproperty-p) and [`[[Get]]`](https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects-get-p-receiver) methods regarding invalid indices:

```js
Object.defineProperty(Int8Array.prototype, -1, {
  get() { throw new Error("-1 getter should be unreachable!"); },
  set(-v) { throw new Error("-1 setter should be unreachable!"); },
});

const tA = new Int8Array(10);
-1 in tA; // => false
tA[-1]; // => undefined
tA[-1] = 1; // => throws Error
```

#1556 aligns `[[Set]]` with other methods regarding canonical numeric indices that are invalid, and is what V8 currently do.
